### ISSUE TITLE
Fails at plugin load on Apple Silicon Macs

### ISSUE THREAD

**Issue by tpayne9** [mcaceresb/stata-shasum#1]
Fails with "Could not load plugin" on Apple Silicon Macs (arm64).  I recompiled openssl and the shasum plugin on my M1 Mac, which is in the arm64 branch of my fork.  

But I don't know how to compile the plugin or make other alterations in a way that both Intel and Apple Silicon Macs can use, so I'm filing this as an issue instead of as a pull request.  

Thanks for providing this command, by the way.  It's been very useful for my data analysis projects!

**Comment by mcaceresb** [comment]
@tpayne9 Finally got a sec to look into this issue; when you compiled, did the plugin fail (i.e. do you get an error with the Stata function)? I recently tested out an update to the plugin I'd been putting off for some time, and in doing so I checked the Linux and Windows versions are working.

However, I don't have OSX hardware, so the best I could do was check that it compiles correctly (which it does, with some minor modifications to the instructions in the README, that are now updated). LMK.

**EDIT:** I mean to ask whether the plugin continued to fail after you re-compiled. One way to test it, by the way, is to run the `tests.do` script from `build`.

**Comment by tpayne9** [comment]
@mcaceresb After recompile, shasum passes your tests.do and all is well.  

Your revised readme's instructions were straightforward and worked well to resolve it.  The only hiccup I had was related to paths in the net install line (my pwd at the time was already stata-shasum/build).

It would be great if a recompile weren't necessary on M1 Macs, but a quick glance at StackOverflow regarding compiling a universal binary of OpenSSL for both M1 and Intel Macs suggests that may be complicated.

Thanks again for your efforts on this command.  My group uses this tool to hash sensitive identifiers in research data, and our solutions before this existed were far worse.

**Comment by mcaceresb** [comment]
> It would be great if a recompile weren't necessary on M1 Macs, but a quick glance at StackOverflow regarding compiling a universal binary of OpenSSL for both M1 and Intel Macs suggests that may be complicated.

@tpayne9 It's not specific to OpenSSL; this seems to be the case now in general with plugins. Compiled code is platform-specific; while the same compiled code typically works in multiple versions of the same OS, for whatever reason the newer OSX chips don't jive with the code I compile on github's OSX servers.  This happens in my other projects as well.

**Comment by mcaceresb** [comment]
@tpayne9 PS: From Stata 16 there's python integration, by the way, so you could side-step the need for a plugin if you're willing to use python. Here's some quick code that does the same thing:

```stata
program pythonhash
    syntax varlist, [GENerate(str) hash(str)]
    python: from sfi import Data
    python: from numbers import Number
    python: from struct import pack
    python: from hashlib import `hash'
    python: array    = Data.get("`varlist'".split(' '))
    python: packnum  = lambda val: pack("d", float(val))
    python: packstr  = lambda val: pack(f'{len(val)}s', bytes(val.encode()))
    python: packval  = lambda val: packnum(val) if isinstance(val, Number) else packstr(val)
    python: packrow  = lambda row: b''.join(packval(val) for val in row)
    python: hashes   = [`hash'(packrow(row)).hexdigest() for row in array]
    python: Data.addVarStr("`generate'", len(hashes[0]))
    python: Data.store("`generate'", None, hashes)
end

sysuse auto, clear
gen byte test = mod(_n, 2)
pythonhash price make test headroom, gen(pymd5) hash(md5)
pythonhash price make test headroom, gen(py256) hash(sha256)
shasum price make test headroom, md5(statamd5) sha256(stata256)
assert pymd5 == statamd5
assert py256 == stata256
```

**Comment by tpayne9** [comment]
@mcaceresb That's **very** useful as well!  Thanks again.  You're more helpful than you know.

My shop is a group of social scientists and analysts... we're not really programmers, although we occasionally do programmer-ish things.  I'm the only one who even has python installed.  

The attractive thing about -shasum- prior to Apple's change in CPU architecture was that we could use commands internal to Stata (net from, net install, etc) without anything external to Stata, and it worked across all operating systems we use.  After the initial install, it's also easy to package -shasum- for use on air-gapped computers, where we work on the most confidential data we have.

Having to recompile -shasum- or install python isn't an insurmountable obstacle, but it is less ideal.  I suspect we'll end up installing python everywhere, since that has other uses and Stata's current integration is pretty good.

**Comment by mcaceresb** [comment]
@tpayne9 I made `pyhash` a function and part of the shasum installation. See `help pyhash` for more; it uses python and is based on my snippet above. Cheers.

**Comment by mcaceresb** [comment]
Should be fixed in 6d6476a8f5885dead26a849e2aff1f8e0f41aaa5
